File: UnitTest.cpp
#include "Solution.hpp"

struct testcase {
    string s;
    vector<vector<int>> shifts;
    string output;
};

class UnitTest {
private:
    Solution obj;
    vector<testcase> testcases;
public:
    UnitTest() {
        testcases = {{"abc", {{0,1,0},{1,2,1},{0,2,1}}, "ace"},
                     {"dztz", {{0,0,0},{1,1,1}}, "catz"}};
    }

    void test() {
        for(int i = 0; i < testcases.size(); ++i) {
            string s = testcases[i].s;
            vector<vector<int>> shifts = testcases[i].shifts;
            string output = testcases[i].output;

            string result = obj.shiftingLetters(s, shifts);
            cout << "Testcase " << i+1 << ": " << ((result == output) ? "passed" : "failed") << endl;
        }
    }
};

int main() {
    UnitTest test;
    test.test();
}

File: UnitTest.py
from Solution import Solution
from timeout_decorator import timeout
import unittest

class UnitTest(unittest.TestCase):
    def setUp(self):
        self.__testcases = {1: ('abc', [[0,1,0],[1,2,1],[0,2,1]], 'ace'),
                            2: ('dztz', [[0,0,0],[1,1,1]], 'catz')}
        self.__obj = Solution()
        return super().setUp()

    @timeout(0.5)
    def test_case_1(self):
        s, shifts, output = self.__testcases[1]
        result = self.__obj.shiftingLetters(s = s, shifts = shifts)
        self.assertIsInstance(result, str)
        self.assertEqual(result, output)

    @timeout(0.5)
    def test_case_2(self):
        s, shifts, output = self.__testcases[2]
        result = self.__obj.shiftingLetters(s = s, shifts = shifts)
        self.assertIsInstance(result, str)
        self.assertEqual(result, output)

if __name__ == '__main__': unittest.main()